prov/mrail/src/mrail_cq.c

```diff
@@ -113,22 +113,145 @@ int mrail_cq_process_buf_recv(struct fi_cq_tagged_entry *comp,
 	return retv;
 }
 
-static int mrail_cq_process_comp_buf_recv(struct util_cq *cq,
-					  struct fi_cq_tagged_entry *comp)
+/* Should only be called while holding a lock */
+static struct mrail_recv *mrail_match_recv(struct mrail_ep *mrail_ep,
+					   struct fi_cq_tagged_entry *comp)
+{
+	struct mrail_hdr *hdr = comp->buf;
+	struct mrail_recv *recv;
+
+	if (hdr->op == ofi_op_msg) {
+		FI_DBG(&mrail_prov, FI_LOG_CQ, "Got MSG op\n");
+		// TODO pass the right address
+		recv = mrail_match_recv_handle_unexp(&mrail_ep->recv_queue, 0,
+						     FI_ADDR_UNSPEC,
+						     (char *)comp, sizeof(*comp),
+						     NULL);
+	} else {
+		assert(hdr->op == ofi_op_tagged);
+		FI_DBG(&mrail_prov, FI_LOG_CQ, "Got TAGGED op\n");
+		recv = mrail_match_recv_handle_unexp(&mrail_ep->trecv_queue,
+						     hdr->tag, FI_ADDR_UNSPEC,
+						     (char *)comp, sizeof(*comp),
+						     NULL);
+	}
+
+	return recv;
+}
+
+static
+struct mrail_ooo_recv *mrail_get_next_recv(struct mrail_peer_info *peer_info)
+{
+	struct slist *queue = &peer_info->ooo_recv_queue;
+	struct mrail_ooo_recv *ooo_recv;
+
+	if (!slist_empty(queue)) {
+		ooo_recv = container_of(queue->head, struct mrail_ooo_recv,
+				entry);
+		if (ooo_recv->seq_no == peer_info->expected_seq_no) {
+			slist_remove_head(queue);
+			peer_info->expected_seq_no++;
+			return ooo_recv;
+		}
+	}
+	return NULL;
+}
+
+static int mrail_process_ooo_recvs(struct mrail_ep *mrail_ep,
+				   struct mrail_peer_info *peer_info)
+{
+	struct mrail_ooo_recv *ooo_recv;
+	struct mrail_recv *recv;
+	int ret;
+
+	ofi_ep_lock_acquire(&mrail_ep->util_ep);
+	ooo_recv = mrail_get_next_recv(peer_info);
+	if (!ooo_recv) {
+		goto exit;
+	}
```

Comment:
This if statement can be dropped.  The while statement does the same check and ends at the same location.